pageserver/src/tenant/mgr.rs

```diff
@@ -244,12 +247,55 @@ pub async fn shutdown_all_tenants() {
         }
     };
 
+    // Set tenant (and its timlines) to Stoppping state.
+    // Since we can only transition into Stopping state after activation is complete,
+    // run it in a JoinSet so all tenants have a chance to stop before we git SIGKILLed.
+    //
+    // Transitioning tenants to Stopping state has a couple of non-obvious side effects:
+    // 1. Lock out any new requests to the tenants.
+    // 2. Signal cancellation to WAL receivers (we wait on it below).
+    // 3. Signal cancellation for othher tenant background loops.
+    // 4. ???
+    //
+    // The waiting for the cancellation is not done uniformly.
+    // We certainly wait for WAL receivers to shut down.
+    // That is necessary so that no new data comes in before the freeze_and_flush.
+    // But the tenant background loops are joined-on in our caller.
+    // It's mesed up.
+    let mut join_set = JoinSet::new();
     let mut tenants_to_freeze_and_flush = Vec::with_capacity(tenants_to_shut_down.len());
     for (_, tenant) in tenants_to_shut_down {
-        if tenant.is_active() {
-            // updates tenant state, forbidding new GC and compaction iterations from starting
-            tenant.set_stopping();
-            tenants_to_freeze_and_flush.push(tenant);
+        join_set.spawn(async move {
+            match tenant.set_stopping().await {
+                Ok(()) => Ok(tenant),
+                Err(e) => Err((tenant, e)),
+            }
+        });
+    }
+    while let Some(res) = join_set.join_next().await {
+        match res {
+            Err(join_error) if join_error.is_cancelled() => {
+                unreachable!("we are not cancelling any of the futures");
+            }
+            Err(join_error) => {
+                // cannot really do anything, as this panic is likely a bug
+                error!("task that calls set_stopping() panicked, don't know which tenant this is, and probably freeze_and_flush won't work anyways: {join_error:#}");
```

Comment:
The hook will have already reported this.

The only possible thing missing is a span around join_set spawned task, which could let the tenant_id be known, but for this site, it doesn't matter.

`set_stopping` does not have `#[instrument]`, I don't have the callers of shutdown_pageserver right now but I am quite certain they are not or can not set a useful span for this panick.

So, I'll patch this to be for example a counter an a single warning, then add the `.instrument(...)` for each of the `join_set.spawn`'d futures.

Reply:
Ok, I'll leave that to you.

Reply:
2cbb593d (fix this conv up) and 745e3d45 (adds `#[instrument]`s)

Reply:
745e3d4 forgot an `use`...

Reply:
[6bee7df](https://github.com/neondatabase/neon/pull/4321/commits/6bee7df00b0cc69fcc3b6fa4b0b53b20dfb84e67) and [71fa6d9](https://github.com/neondatabase/neon/pull/4321/commits/71fa6d924ee148c9d199729ad03de2c5ceb3d80e) after a bit of touchup.